pkg/domain/domain.go

```diff
@@ -1863,52 +1929,39 @@ func (do *Domain) LoadPrivilegeLoop(sctx sessionctx.Context) error {
 		}()
 		defer util.Recover(metrics.LabelDomain, "loadPrivilegeInLoop", nil, false)
 
-		var (
-			closeChanCnt int
-			resp         clientv3.WatchResponse
-		)
+		var count int
 		for {
-			var ok bool
+			var event PrivilegeEvent
 			select {
 			case <-do.exit:
 				return
-			case resp, ok = <-watchCh:
-				if !ok {
-					logutil.BgLogger().Error("load privilege loop watch channel closed")
-					watchCh = do.etcdClient.Watch(do.ctx, privilegeKey)
-					closeChanCnt++
-					if closeChanCnt > 10 {
-						time.Sleep(time.Duration(closeChanCnt) * time.Second)
-					}
-					continue
-				}
-				closeChanCnt = 0
-
-				// Skip the event from this TiDB-Server
-				eventCnt := len(resp.Events)
-				for _, event := range resp.Events {
-					if event.Kv != nil {
-						val := event.Kv.Value
-						if len(val) > 0 {
-							var tmp PrivilegeEvent
-							err := json.Unmarshal(val, &tmp)
-							if err != nil {
-								logutil.BgLogger().Warn("decode PrivilegeEvent unmarshal fail", zap.Error(err))
-								continue
-							}
-							if do.serverID != 0 && do.serverID == tmp.ServerID {
-								eventCnt--
-							}
+			case resp, ok := <-watchCh:
+				if ok {
+					count = 0
+					event = do.decodePrivilegeEvent(resp)
+					event = do.batchReadMoreData(watchCh, event)
+				} else {
+					if do.ctx.Err() == nil {
+						logutil.BgLogger().Warn("load privilege loop watch channel closed")
+						watchCh = do.etcdClient.Watch(do.ctx, privilegeKey)
+						count++
+						if count > 10 {
+							time.Sleep(time.Duration(count) * time.Second)
 						}
+						continue
 					}
 				}
-				if eventCnt == 0 {
-					continue
-				}
 			case <-time.After(duration):
+				event.All = true
+				event = do.batchReadMoreData(watchCh, event)
```

Comment:
event.All = true, there is no need to batch read more, it forces a full reload...
Is master the same here?

Reply:
Yes, master is the same here, since https://github.com/pingcap/tidb/pull/61147/files